trpcultivate_phenotypes/src/TripalCultivateValidator/ValidatorTraits/Headers.php

```diff
@@ -0,0 +1,213 @@
+    // in designated type array of the context.
+    $context_headers = [];
+
+    foreach($headers as $index => $header) {
+      // Header element key and value check.
+      foreach($required_header_keys as $key) {
+        // Key is not set.
+        if (!isset($header[ $key ])) {
+          throw new \Exception('Headers Trait requires the header key: ' . $key . ' when defining headers.');
+        }
+        
+        // Key is set but value is empty.
+        if (empty(trim($header[ $key ]))) {
+          throw new \Exception('Headers Trait requires the header key: ' . $key . ' to be have a value.');
+        }
+
+        // Type value is one of valid types.
+        if ($key == 'type' && !in_array($header[ $key ], $this->types)) {
+          $str_types = implode(', ', $this->types);
+          throw new \Exception('Headers Trait requires the header key: ' . $key . ' value to be one of [' . $str_types . '].');
+        }
+      }
+
+      // With the header type already verified to be one of the valid types, 
+      // push the header into the designated type temporary context array.
+      $context_headers[ $header['type'] ][ $index ] = $header['name'];
+    }
+    
+    // Set each header type context array.
+    foreach($this->types as $type) {  
+      $this->context[ $this->context_key ][ $type ] = $context_headers[ $type ] ?? [];
+    }
+  }
+  
+  /**
+   * Get required headers.
+   * 
+   * @return array
+   *   All headers of type required, keyed by the index (order) from
+   *   the headers array and header name as the value.
+   * 
+   *   The key required in the context headers array set by the setter method.
+   *   NOTE: the headers array is zero-based index.
+   * 
+   * @throws \Exception
+   *  - If the 'headers' key does not exists in the context array
+   *    (ie. the headers element has NOT been set).
+   */
+  public function getRequiredHeaders() {
+    
+    $type_key = 'required';
+
+    if (array_key_exists($this->context_key, $this->context)) {
+      return $this->context[ $this->context_key ][ $type_key ];
+    }
+    else {
+      throw new \Exception('Cannot retrieve ' . $type_key . ' headers from the context array as one has not been set by setHeaders() method.');
+    }
+  }
+  
+  /**
+   * Get optional headers.
+   * 
+   * @return array
+   *   All headers of type optional, keyed by the index (order) from
+   *   the headers array and header name as the value.
+   * 
+   *   The key optional in the context headers array set by the setter method.
+   *   NOTE: the headers array is zero-based index.
+   * 
+   * @throws \Exception
+   *  - If the 'headers' key does not exists in the context array
+   *    (ie. the headers element has NOT been set).
+   */
+  public function getOptionalHeaders() {
+    
+    $type_key = 'optional';
+
+    if (array_key_exists($this->context_key, $this->context)) {
+      return $this->context[ $this->context_key ][ $type_key ];
```

Comment:
Again, `$this->context['headers']['optional']` needs to be checked if it exists separately.